apps/oracles/stock-price-feeds/src/providers/twelvedata.rs

```diff
@@ -0,0 +1,86 @@
+use std::collections::HashMap;
+
+use anyhow::{Error, Result};
+use futures::{future::LocalBoxFuture, FutureExt};
+
+use serde::Deserialize;
+
+use blocksense_sdk::{
+    http::http_get_json,
+    traits::prices_fetcher::{PairPriceData, PricePoint, PricesFetcher},
+};
+
+#[derive(Default, Debug, Clone, PartialEq, Deserialize)]
+pub struct PriceData {
+    pub symbol: String,
+    pub close: Option<String>,
+    pub volume: Option<String>,
+    pub previous_close: Option<String>,
+}
+
+type TwelveDataResponse = HashMap<String, PriceData>;
+
+pub struct TwelveDataPriceFetcher<'a> {
+    pub symbols: &'a [String],
+    api_key: Option<&'a str>,
+}
+
+impl<'a> PricesFetcher<'a> for TwelveDataPriceFetcher<'a> {
+    const NAME: &'static str = "twelvedata";
+
+    fn new(symbols: &'a [String], api_key: Option<&'a str>) -> Self {
+        Self { symbols, api_key }
+    }
+
+    fn fetch(&self) -> LocalBoxFuture<Result<PairPriceData>> {
+        async {
+            let api_key = match self.api_key {
+                Some(key) => key,
+                None => {
+                    return Err(Error::msg("API key is required for TwelveData"));
+                }
+            };
+
+            let all_symbols = self.symbols.join(",");
+
+            let response = http_get_json::<TwelveDataResponse>(
+                "https://api.twelvedata.com/quote",
+                Some(&[("symbol", &all_symbols)]),
+                Some(&[("Authorization", format!("apikey {api_key}").as_str())]),
+            )
+            .await?;
+
+            let results = response
+                .values()
+                .filter_map(|value| {
+                    let price = value.close.clone().or(value.previous_close.clone());
+                    let volume = value.volume.clone();
+
+                    match (price.clone(), volume.clone()) {
+                        (Some(price), Some(volume)) => {
+                            let price = price.parse::<f64>().unwrap_or(0.0);
+                            let volume = volume.parse::<f64>().unwrap_or(0.0);
+                            Some((value.symbol.clone(), PricePoint { price, volume }))
+                        }
+                        _ => {
+                            eprintln!(
+                                "[TwelveData] Skipping symbol {}: missing {}{}{}",
+                                value.symbol,
+                                if price.is_none() { "price" } else { "" },
+                                if price.is_none() && volume.is_none() {
+                                    " and "
+                                } else {
+                                    ""
+                                },
+                                if volume.is_none() { "volume" } else { "" },
```

Comment:
nit: It seems we use this log in most of the providers' fetchers. In the future, we can export it to `apps/oracles/stock-price-feeds/src/utils.rs`